fastfield_codecs/src/optional_column.rs

```diff
@@ -0,0 +1,120 @@
+use std::ops::{Range, RangeInclusive};
+use std::sync::Arc;
+
+use crate::Column;
+
+/// `OptionalColumn` provides columnar access on a field.
+pub trait OptionalColumn<T: PartialOrd = u64>: Send + Sync {
+    /// Return the value associated with the given idx.
+    ///
+    /// This accessor should return as fast as possible.
+    ///
+    /// # Panics
+    ///
+    /// May panic if `idx` is greater than the column length.
+    fn get_val(&self, idx: u32) -> Option<T>;
+
+    /// Fills an output buffer with the fast field values
+    /// associated with the `DocId` going from
+    /// `start` to `start + output.len()`.
+    ///
+    /// # Panics
+    ///
+    /// Must panic if `start + output.len()` is greater than
+    /// the segment's `maxdoc`.
+
+    fn get_range(&self, start: u64, output: &mut [Option<T>]) {
+        for (out, idx) in output.iter_mut().zip(start..) {
+            *out = self.get_val(idx as u32);
+        }
+    }
+
+    /// Return the positions of values which are in the provided range.
+    fn get_docids_for_value_range(
+        &self,
+        value_range: RangeInclusive<T>,
+        doc_id_range: Range<u32>,
+        positions: &mut Vec<u32>,
+    ) {
+        let doc_id_range = doc_id_range.start..doc_id_range.end.min(self.num_vals());
+
+        for idx in doc_id_range.start..doc_id_range.end {
+            let val = self.get_val(idx);
+            if let Some(val) = val {
+                if value_range.contains(&val) {
+                    positions.push(idx);
+                }
+            }
+        }
+    }
+
+    /// Returns the minimum value for this fast field.
+    ///
+    /// This min_value may not be exact.
+    /// For instance, the min value does not take in account of possible
+    /// deleted document. All values are however guaranteed to be higher than
+    /// `.min_value()`.
+    fn min_value(&self) -> Option<T>;
```

Comment:
I wonder if we should make it `T` and force this trait to have at least one value?
same for max_value.

Reply:
Then we would need to return `Option<dyn OptionalColumn>`. It's a valid alternative, but the handling may be more complex